### PR TITLE
Use function pointers in more places

### PR DESCRIPTION
**Pull request by jkotas**

### CODE REVIEW

**Comment by jkotas** [comment]
I know that these will need minor updates once we get Roslyn with the final syntax. Given how effective using the function pointers in our own libraries was in shaking corner case bugs, I think it is worth it.

File: src/coreclr/src/System.Private.CoreLib/src/System/Runtime/InteropServices/Marshal.CoreCLR.cs
Comment (AaronRobinsonMSFT) [inline src/coreclr/src/System.Private.CoreLib/src/System/Runtime/InteropServices/Marshal.CoreCLR.cs:712]: This is soon to be invalid syntax. It might make sense to wait for the update?

/cc @333fred @jkoritzinsky

**Review by AaronRobinsonMSFT — COMMENTED** [review]
LGTM as long as we are okay updating this when we take a Roslyn compiler update. Seems like it would be less annoying to wait, but I love seeing the deleted code.

**Comment by AaronRobinsonMSFT** [comment]
> I know that these will need minor updates once we get Roslyn with the final syntax. Given how effective using the function pointers in our own libraries was in shaking corner case bugs, I think it is worth it.

Okay. What is the process for consuming Roslyn updates? I would hate for some CI to be blocked on this and people now knowing why for too long. Can one of us be proactive in shepherding that change?

File: src/coreclr/src/System.Private.CoreLib/src/System/Runtime/InteropServices/Marshal.CoreCLR.cs
Comment (AaronRobinsonMSFT) [reply src/coreclr/src/System.Private.CoreLib/src/System/Runtime/InteropServices/Marshal.CoreCLR.cs:712]: Acknowledged at https://github.com/dotnet/runtime/pull/39752#issuecomment-662181763.

**Comment by jkotas** [comment]
> What is the process for consuming Roslyn updates? 

It shows up as Arcade SDK update by default. It is also possible to pick up a new Roslyn explicitly, without waiting for the Arcade SDK update, for example: https://github.com/dotnet/runtime/pull/36104/files#diff-8b8f08ffbf7b863fb3700c1718eeb4cbR22

**Comment by AaronRobinsonMSFT** [comment]
@jkotas Sounds good. I would suggest then either you or I proactively handle moving to the latest Roslyn when it gets checked in.

@333fred if you could let @jkotas or myself know when the the updates are in Roslyn main that would be helpful.

**Review by AaronRobinsonMSFT — APPROVED** [review]

**Comment by jkoritzinsky** [comment]
One more place to use function pointers is in the tail call dispatcher. It's currently an IL stub because it needed calli support. It should be convertable to C# pretty easily now.

**Comment by jkotas** [comment]
> One more place to use function pointers is in the tail call dispatcher

That one is fixed already :-)

**Comment by AaronRobinsonMSFT** [comment]
@jkoritzinsky I was looking for tail call change too: https://github.com/dotnet/runtime/pull/38938

File: src/libraries/Microsoft.CSharp/src/Microsoft/CSharp/RuntimeBinder/ComInterop/ComRuntimeHelpers.cs
Comment (333fred) [inline src/libraries/Microsoft.CSharp/src/Microsoft/CSharp/RuntimeBinder/ComInterop/ComRuntimeHelpers.cs:295]: This might be one of the most terrifying lines of code I've ever seen.

**Pull request merged**
